wolfssl: disable AES-NI by default for x86_64

cotequeiroz opened the pull request:
WolfSSL is crashing with an illegal opcode in some x86_64 CPUs that have
AES instructions but lack other extensions that are used by WolfSSL
when AES-NI is enabled.

Disable the option by default for now until the issue is properly fixed.
People can enable them in a custom build if they are sure it will work
for them.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

---

This was reported in https://github.com/openwrt/openwrt/pull/9738#issuecomment-1162041335

cotequeiroz committed:
wolfssl: disable AES-NI by default for x86_64

WolfSSL is crashing with an illegal opcode in some x86_64 CPUs that have
AES instructions but lack other extensions that are used by WolfSSL
when AES-NI is enabled.

Disable the option by default for now until the issue is properly fixed.
People can enable them in a custom build if they are sure it will work
for them.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

cotequeiroz commented:
I may be the one to blame for this.  I've enabled all intel asm supeedups, mistakenly thinking they would be detected at runtime like AES-NI is.  

What I can do is break it into two options: use `--enable-aesni` as standard, and add another option for custom builds (`--enable-intelasm`), with the big fat warning, for the full Intel asm support.  We need to have this tested fist though, so it may be best to merge this with the default `off`, and then use aesni if it is safe.

neheb commented:
this was pointed out on IRC as well.

Ansuel commented:
Also i think jow tested without the msee3 config flag and same problem

hauke commented:
You can reproduce it with this configuration:
````
CONFIG_TARGET_x86=y
CONFIG_TARGET_x86_64=y
CONFIG_TARGET_x86_64_DEVICE_generic=y
````
Then you can start qemu like this: `./scripts/qemustart x86 64`

hauke commented:
Thank you for the patch, I added it to the master branch in 0bd536723303ccd178e289690d073740c928bb34.

cotequeiroz commented:
This appears to be generation of SSE4.1 code by gcc, for example this code in [wolfcrypt/src/tfm.c](https://github.com/wolfSSL/wolfssl/blob/b87b255d5280604f93bacb2d6170bb7c30169af3/wolfcrypt/src/tfm.c#L230-L243):
``` C
/* from wolfssl/wolfcrypt/tfm.h */
/* a FP type */
typedef struct fp_int {
    int      used;
    int      sign;
#if defined(ALT_ECC_SIZE) || defined(HAVE_WOLF_BIGINT)
    int      size;
#endif
    fp_digit dp[FP_SIZE];

#ifdef HAVE_WOLF_BIGINT
    struct WC_BIGINT raw; /* unsigned binary (big endian) */
#endif
} fp_int;

/* from wolfssl/woflcrypt/wolfmath.h */
#define MAX(x,y) ((x)>(y)?(x):(y))

int fp_mul(fp_int *A, fp_int *B, fp_int *C)
{
    int   ret = 0;
    int   y, yy, oldused;
/*...*/
    y  = MAX(A->used, B->used);
```
results in 
``` asm
pminsd %xmm1,%xmm0
```
`pminsd` is available with SSE 4.1.

I used `wolfssl-benchmark` to trigger the error.

The crash above was after changing `-Os` to `-O2`.  The crash with the vanilla library was in a different place, but it was harder to see what was going on.
**wolfcrypt/src/pwdbased.c:**
``` C
ret = wc_HmacSetKey(hmac, hashType, passwd, pLen);
```
``` asm
pmaxsd (%rsp),%xmm0
```
For completeness, here's the config I used:
```
CONFIG_TARGET_x86=y
CONFIG_TARGET_x86_64=y
CONFIG_TARGET_x86_64_DEVICE_generic=y
CONFIG_DEVEL=y
CONFIG_DEBUG=y
CONFIG_NO_STRIP=y
CONFIG_PACKAGE_gdb=y
CONFIG_PACKAGE_libgmp=y
CONFIG_PACKAGE_libncurses=y
CONFIG_PACKAGE_libreadline=y
CONFIG_PACKAGE_libwolfssl-benchmark=y
CONFIG_PACKAGE_terminfo=y
CONFIG_PACKAGE_zlib=y
CONFIG_TARGET_OPTIMIZATION="-O2 -pipe"
CONFIG_TARGET_OPTIONS=y
CONFIG_TARGET_ROOTFS_PARTSIZE=248
# CONFIG_USE_SSTRIP is not set
CONFIG_WOLFSSL_HAS_CPU_CRYPTO=y
# CONFIG_WOLFSSL_HAS_NO_HW is not set
```

I've hacked a way to make AESNI work.
- removed `-maes -mpclmul -sse4` flags from `configure.ac`
- run `./configure` with `--enable-aesni` instead of `--enable-intelasm`
- added `-maes` to `CFLAGS` only for `wolfcrypt/src/aes.c`

It worked on qemu, and I confirmed it uses AESNI in the host, so I'm confident it will work. It may be by accident, as shown by `-O2` going further than `-Os` through the benchmark, so it does need to be tested. I'll open a RFT PR when able, and then work my way into submitting a proper fix upstream.
